httpclient5/src/main/java/org/apache/hc/client5/http/protocol/RequestTraceInterceptor.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * <p><strong>Responsibilities:</strong></p>
+ * <ul>
+ *   <li>Sanitizes TRACE requests by removing sensitive headers such as {@code Authorization} and {@code Cookie}.</li>
+ *   <li>Ensures that TRACE requests do not contain a request body, throwing a {@link ProtocolException} if a body is present.</li>
+ * </ul>
+ *
+ * <p><strong>Thread Safety:</strong> This class is stateless and therefore thread-safe, as indicated by its {@code ThreadingBehavior.STATELESS} annotation.</p>
+ *
+ * <p><strong>Interceptor Behavior:</strong></p>
+ * <ul>
+ *   <li>If the HTTP method is TRACE, the interceptor removes any {@code Authorization} and {@code Cookie} headers to prevent sensitive data leakage.</li>
+ *   <li>If a TRACE request contains a body, a {@link ProtocolException} is thrown.</li>
+ * </ul>
+ *
+ * @version 5.4
+ * @see HttpRequestInterceptor
+ * @see HttpException
+ * @see IOException
+ * @see ProtocolException
+ * @see Method#TRACE
+ * @see HttpHeaders#AUTHORIZATION
+ * @see HttpHeaders#COOKIE
+ */
+@Contract(threading = ThreadingBehavior.STATELESS)
+public class RequestTraceInterceptor implements HttpRequestInterceptor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RequestTraceInterceptor.class);
+
+    /**
+     * Singleton instance of {@link RequestTraceInterceptor}.
+     */
+    public static final HttpRequestInterceptor INSTANCE = new RequestTraceInterceptor();
+
+    /**
+     * Default constructor.
+     */
+    public RequestTraceInterceptor() {
+        super();
+    }
+
+    /**
+     * Processes an incoming HTTP request. If the request is of type TRACE, it performs the following actions:
+     * <ul>
+     *   <li>Removes the {@code Authorization} header to prevent sensitive data leakage.</li>
+     *   <li>Removes the {@code Cookie} header to prevent sensitive data leakage.</li>
+     *   <li>Throws a {@link ProtocolException} if the request contains a body.</li>
+     * </ul>
+     *
+     * @param request The incoming HTTP request. Cannot be {@code null}.
+     * @param entity  Details of the request entity. Can be {@code null}.
+     * @param context The HTTP context.
+     * @throws HttpException If a protocol error occurs.
+     * @throws IOException   If an I/O error occurs.
+     */
+    @Override
+    public void process(final HttpRequest request, final EntityDetails entity, final HttpContext context)
+            throws HttpException, IOException {
+
+        Args.notNull(request, "HTTP request");
+        Args.notNull(context, "HTTP context");
+
+        // Check if the request method is TRACE
+        if (Method.TRACE.isSame(request.getMethod())) {
+
+            //A client MUST NOT send content in a TRACE request.
+            if (entity != null) {
+                throw new ProtocolException("TRACE request MUST NOT contain a request body.");
+            }
+
+            // Check for sensitive headers
+            final Header authHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
```

Comment:
@arturobernalg I really dislike the idea of removing headers explicitly set by the caller. It feels wrong. Would you be OK with throwing a protocol exception instead?

Reply:
@ok2c 

Fair enough. changed